text/0003-font-types.md

```diff
@@ -0,0 +1,408 @@
+of font tables and types, leveraging Rust’s procedural macro system. This
+library is intended to be an opinion-free transparent representation of the types
+in the specification, to serve as a foundational component for anything that
+needs to read or write font files. An additional goal is to provide a common set
+of types to represent common data formats (tags, glyph identifiers, etc) that
+can be used throughout the broader ecosystem.
+
+
+# Motivation
+
+This is intended to answer some of the questions outlined in this repository's
+README. In particular, we aim to determine whether the following goals are feasible:
+
+- using macros to describe the shape of tables
+-copy generating both zero-allocation read-only table definitions as well as owned,
+  mutable definitions (for compilation)
+- suitable for prototyping & extensibility, i.e. it should be easy to add a new
+  table type for a project without needing to patch the core crate or maintain a
+  fork.
+
+# Design
+
+The particulars of the design will be determined through this exploration, but a
+few core ideas are clear. For the read-only case, the design is largely borrowed
+directly from [pinot][], with the main difference being using macros for code
+generation, and that approach has been successful for [fonttools-rs][].
+
+## conversion to and from big-endian bytes
+
+Values in the binary tables are generally big-endian and may not be aligned to
+word boundaries, and so have to be converted to the appropriate scalars when
+read (In this sense a font library can never be exactly 'zero-copy', although it
+can be zero-allocation, and support random access.)
+
+Scalar values that can be decoded (and later encoded) from raw font bytes will
+be represented by a trait, something like:
+
+```rust
+trait FromFontBytes: Sized {
+    /// Number of bytes required to represent this type
+    const SIZE: usize = std::mem::size_of::<Self>();
+
+    fn from_be_bytes(bytes: &[u8]) -> Result<Self, Error>;
+}
+
+impl FromFontBytes for u8 {
+    fn from_be_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        bytes.first().copied().ok_or(Error::Eof)
+    }
+}
+```
+
+This trait can then also be derived for types that are entirely composed of
+scalar values that implement this trait:
+
+```rust
+#[derive(Debug, Clone, FromFontBytes)]
+#[repr(C)] // ensure field layout matches declaration order
+struct AxisValueMap {
+    from_coordinate: F2dot14,
+    to_coordinate: F2dot14,
+}
+```
+
+## zero-allocation views into bytes & random access
+
+### `Buffer`
+
+When reading font data, the library will perform no allocation, and instead
+provide 'views' for accessing the different tables and their constituent
+elements. This API will be based around a `Buffer` type, which is an abstraction
+over some sequence of bytes. The actual storage of the those bytes is hidden
+from the user, but it might be some bytes loaded from disk, a memory-mapped
+file, or an owned buffer.
```

Comment:
I interpret this to deliberately parallel the HB buffer strategy. If so, please explicitly note as much and perhaps link to def or docs for hb_buffer_t? If not perhaps clarify that :)

Reply:
In this case, I believe Buffer is similar to the same type in pinot (see: https://github.com/dfrg/pinot/blob/7001393e7ef824ca99a7ed60a7592a9950a88230/src/parse/mod.rs#L16) which wraps the font data and allows reading values of the basic types at arbitrary offsets.

Reply:
In HB that's a `hb_blob_t`.